README
== Description
  A Ruby interface for gathering process information.

== Prerequisites
* RSpec 3.x (development only)

== Supported Platforms
* Windows 2000 or later
* Linux 2.6+
* FreeBSD
* Solaris 8+
* OS X 10.7+
* AIX 5.3+

== Installation
  gem install sys-proctable

  For version 1.1.5 or earlier, you may need to specify a platform in some cases. For example:

  gem install sys-proctable --platform mswin32 # Windows
  gem install sys-proctable --platform sunos   # Solaris
  gem install sys-proctable --platform linux   # Linux
  gem install sys-proctable --platform freebsd # FreeBSD
  gem install sys-proctable --platform darwin  # OS X

== Synopsis
  require 'sys/proctable'
  include Sys

  # Everything
  ProcTable.ps{ |p|
    puts p.pid.to_s
    puts p.comm
    # ...
  }

  # Just one process
  s = ProcTable.ps(pid: 2123)
  puts s.pid.to_s
  puts s.comm
  # ...

  # Return the results as an array of ProcTableStructs
  a = ProcTable.ps
  a.each do |p|
    puts p.pid
    # ...
  end

== Notes 
  Various platforms support different options. Mostly this is to let you
  skip the collection of certain bits of information in order to improve
  speed and/or reduce memory. For example on Linux you can do this to
  skip the collection of smaps information:

    Sys::ProcTable.ps(smaps: false)

  Windows users may send a host name to get process information from a
  different host. This relies on the WMI service running.

    Sys::ProcTable.ps(host: some_host)

== Known Issues
=== FreeBSD
  A kvm interface is used. That means the owner of the process using the
  sys-proctable library needs to be a member of the kvm group (or root).

=== Bundler
  For version 1.1.5 or earlier, Bundler seems to have trouble installing the
  proper gem because of the platform specific gem names. To deal with that,
  run this command first:

    bundle config specific_platform true

  You should not have to do this for version 1.2.0 or later.

=== Solaris
  The cmdline member on Solaris is limited to 80 characters unless you (or
  your program) own the process. This is a Solaris design flaw/feature.

=== OS X
  The libproc interface is used. That means you will only get list of
  processes that you have access to. To get a full listing, run as root.

== Future Plans
  Support for Solaris will probably be dropped in the next major release.

== Acknowledgements
  This library was originally based on the Perl module Proc::ProcessTable
  by Dan Urist. Many ideas, as well as large chunks of code, were taken
  from his work. So, a big THANK YOU goes out to Dan Urist.

  A big thanks also goes out to Mike Hall who was very helpful with ideas,
  logic and testing.

  Thanks also go to Sean Chittenden for providing an account on one of his
  FreeBSD machines. This is how the FreeBSD support was (initially) added.

  Thanks go to James Hranicky for providing a patch that grabs name, eid,
  euid, gid and guid info in the Linux version, along with some general
  debugging help.
   
  Thanks go to David Felstead for the original OS X code. Thanks also go
  to Matthias Zirnstein for adding the original cmdline support for OS X.

  Finally I'd like to thank all the folks who have submitted bug reports
  and/or patches.

== Help Wanted
  I do not have access to all platforms. If your platform is not supported
  then you will need to either submit a patch or give me a remote account
  on a box with a compiler so that I can write the code.

== More documentation
  See the documentation under the 'doc' directory for more information,
  including platform specific notes and issues.

== License
  Apache 2.0
    
== Copyright
  (C) 2003-2018 Daniel J. Berger
  All Rights Reserved.

== Author
  Daniel J. Berger


CHANGES
== 1.2.3 - 17-Mar-2020
* Properly include a copy of Apache-2.0 in LICENSE file as part of library.

== 1.2.2 - 12-Aug-2019
* Added compatibility for JRuby on Mac which apparently doesn't define a
  read_uint64 method, nor a CharArray. Thanks go to Adithya Pentela for
  the spot.
* Refactored the specs a bit, adding a spec_helper.rb. This was mainly
  for JRuby compatibility.
* Updates to the travis.yml file for both Ruby and JRuby versions.
* Removed the doc directory and all the files it contained. It was
  redundant, and some of the information was wrong. All of the
  documentation you need is on the wiki.
* Cleaned up the MANIFEST file, updated the gemspec, and fixed the
  license name, which was missing a hyphen.

== 1.2.1 - 8-Jun-2018
* The code for OSX is now more efficient when a pid argument is provided.
  Thanks go to Nick LaMuro for his efforts.
* Added metadata to the gemspec.
* Switched the README (now README.md) to markdown format. Thanks go to
  Tim Meusel for the update.
* Updated the cert. Should be good for ten years now.

== 1.2.0 - 20-Feb-2018
* There has been an API change. The ProcTable.ps method now uses keyword
  arguments. The 'pid' option is universal, the rest depend on the platform.
  As part of this change, support for Ruby < 2.0 has been dropped.
* Support for HP-UX has been dropped, both because it was the last remaining
  platform that still used C code, and because it's basically a dead platform.
* There are no more platform-specific gems. There is now a single gem that
  will load the appropriate code based on your host's operating system.
* The tests were switched from test-unit to rspec.
* Lots of Rakefile updates based on the above changes.
* The cert was updated.

== 1.1.5 - 10-Aug-2017
* Fixed a warning that cropped up in Ruby 2.4.x because I was type checking
  against Fixnum. Those have been replaced with Numeric.

== 1.1.4 - 30-Mar-2017
* Fixed a potential issue where OSX could return a struct with invalid data.
* Set the default field to :pid for OSX when using Sys::Top.
* Fixed a duplicate test warning for OSX.

== 1.1.3 - 28-Sep-2016
* Fixed an issue on OSX where the value returned for argc is invalid.

== 1.1.2 - 18-Sep-2016
* Fixed cmdline parsing and handling for OSX. Thanks go to Ben Mathwig for
  the spot and the patch.

== 1.1.1 - 30-Jun-2016
* Added thread information for OS X.
* Fixed VERSION constant for HP-UX.

== 1.1.0 - 27-Jun-2016
* License was changed to Apache 2.0.
* The OS X version now requires OS X 10.7 or later.
* Scrapped the C implementation for OS X, and replaced it with a libproc wrapper
  using FFI, so it is now pure Ruby. Note that some new struct members have been
  added, while others have been dropped. This includes pctcpu, so Sys::Top does
  not work at the moment.
* Rakefile updates to accomodate the changes for OS X.
* Some test suite updates and refactoring.

== 1.0.0 - 11-Jan-2016
* Added smaps information for Linux. Thanks go to Joe Rafaniello for the patch.
* This really should not have been a major release, sorry.

== 0.9.9 - 8-Nov-2015
* Added support for cgroups on Linux. Thanks go to Dennis Günnewig for the patch.
* Added a sys-proctable.rb file for convenience.
* This gem is now signed.
* Gem related tasks in the Rakefile now assume Rubygems 2.x.

== 0.9.8 - 18-Apr-2015
* Fixed a bug in the gemspec. Last version got yanked.

== 0.9.7 - 18-Apr-2015
* Fixed a bug in the OSX code that could cause the ps method to fail. Thanks
  go to Koshevoy Anton for the spot.
* Some internal refactoring of version handling. Now all platforms use a
  single version file.
* Replaced vm_size_t and lwpid_t with universal data types on FreeBSD because
  FFI on FreeBSD 10 no longer understands them. Thanks go to Mike Owens for
  the spot.

== 0.9.6 - 24-Feb-2015
* Added NLWP field on Linux. Thanks go to Rich Chatterton for the patch.

== 0.9.5 - 10-Feb-2015
* Significant cleanup and memory reduction of the OSX code. Thanks go
  to Ivan (toy) for the patches.
* Skip over /proc/<file>/status if unreadable on Linux. Thanks go to Jianjun
  Mao for the patch.

== 0.9.4 - 4-Mar-2014
* Added support for AIX 5.3 or later courtesy of Rick Ohnemus.
* The Solaris version now uses FFI structs instead of a packed array.
  It solved issues with 64-bit versions of Ruby and it's self-documenting.
* The FreeBSD version has been converted to use FFI. In addition, additional
  struct members have been added, and members that previously returned nil
  now return meaningful data.
* Support for NetBSD and OpenBSD has been temporarily dropped. Considering
  that the C code did not build on those platforms anyway, I doubt most of
  you will notice. Patches for those platforms are welcome, but only using FFI.

== 0.9.3 - 17-Mar-2013
* Fixed a bug on OSX where a long command string arg could cause
  a segfault. Thanks go to Nathaniel Bibler for the spot.
* Changed the gem platform from mingw to mingw32 for Windows.

== 0.9.2 - 8-Oct-2012
* Added cmdline support for OS X. Thanks go to Matthias Zirnstein for
  the patch.
* Warning cleanup for 1.9.
* Updated the gem platform handling. Replaced the borked string approach
  with a two element array for Gem::Platform.new.
* MS date strings are now parse with DateTime instead of Date.

== 0.9.1 - 3-Aug-2011
* Added the pctcpu and pctmem members for Linux.
* Added Errno::ESRCH to a rescue clause on Linux that fixed a bug
  where a missing entry wasn't being skipped when run as root. Thanks
  go to Austin Ziegler for the spot and patch.
* Fixed a build warning for Darwin.
* Updates to the test suite for both Darwin and Linux.
* Added an explicit type check for BSD for pids.
* Added nicer error messages for BSD if kvm_open fails.
* Added .core files to the clean task.
* Altered the platform settings in the Rakefile so that generated gems
  use 'universal' platform architectures for any particular operating system.

== 0.9.0 - 14-Oct-2009
* Changed the license to Artistic 2.0.

ext/hpux/sys/proctable.c
/**********************************************************************
* proctable.c
*
* HP-UX specific code for the Ruby ps extension. Some code has been
* copied directly from Dan Urist's Proc::ProcessTable Perl module.
*
* Author: Daniel Berger
**********************************************************************/
#include "ruby.h"

#define _PSTAT64

#ifdef __cplusplus
extern "C"
{
#endif

#include <sys/param.h>
#include <sys/pstat.h>
#include <devnm.h>

#define BURST 30 /* pstat structs to get per syscall */

/* Process state strings */
#define SLEEP  "sleep"
#define RUN    "run"
#define STOP   "stop"
#define ZOMBIE "zombie"
#define IDLE   "idle"
#define OTHER  "other"

VALUE sProcStruct, cProcError;

/*****************************************************************************
 * This array's only purpose is for the 'fields()' class method.  If there's
 * a way to report fields out of our defined struct (sProc), please tell me
 * how!
 *****************************************************************************/
char *fields[] = {
   "comm",
   "uid",
   "pid",
   "ppid",
   "dsize",
   "tsize",
   "ssize",
   "nice",
   "ttydev",
   "pgrp",
   "pri",
   "addr",
   "cpu",
   "utime",
   "stime",
   "start",
   "flag",
   "stat",
   "wchan",
   "procnum",
   "cmd",
   "cmdline",
   "time",
   "cpticks",
   "cptickstotal",
   "fss",
   "pctcpu",
   "rssize",
   "suid",
   "shmsize",
   "mmsize",
   "usize",
   "iosize",
   "vtsize",
   "vdsize",
   "vssize",
   "vshmsize",
   "vmmsize",
   "vusize",
   "viosize",
   "minorfaults",
   "majorfaults",
   "nswap",
   "nsignals",
   "msgrcv",
   "msgsnd",
   "maxrss",
   "sid",
   "schedpolicy",
   "ticksleft",
   "euid",
   "egid",
   "gid",
   "sgid"
};


/*
 * Private method that converts a psinfo struct into a Ruby struct.
 */
static VALUE proctable_getprocstruct(struct pst_status *p)
{
   char state[10];
   char flag[12];
   char ttydev[MAXPATHLEN+1];
   VALUE v_tty, v_struct;

   switch( p->pst_stat )
   {
      case PS_SLEEP:
         strcpy(state,SLEEP);
         break;
      case PS_RUN:
         strcpy(state,RUN);
         break;
      case PS_STOP:
         strcpy(state,STOP);
         break;
      case PS_ZOMBIE:
         strcpy(state,ZOMBIE);
         break;
      case PS_IDLE:
         strcpy(state,IDLE);
         break;
      case PS_OTHER:
         strcpy(state,OTHER);
         break;
      default:
         strcpy(state,"unknown");
   }

   /* If the major number is -1, there is no associated tty */
   if(p->pst_term.psd_major != -1){
      devnm(
         S_IFCHR,
         (dev_t)((p->pst_term.psd_major << 24) | p->pst_term.psd_minor),
         ttydev,
         sizeof(ttydev),
         1
      );
      v_tty = rb_str_new2(ttydev);
   }
   else{
      v_tty = rb_str_new2("");
   }

   v_struct = rb_struct_new(sProcStruct,
      rb_str_new2(p->pst_ucomm),
      INT2NUM(p->pst_uid),
      INT2NUM(p->pst_pid),
      INT2NUM(p->pst_ppid),
      INT2NUM(p->pst_dsize),
      INT2NUM(p->pst_tsize),
      INT2NUM(p->pst_ssize),
      INT2NUM(p->pst_nice),
      v_tty,
      INT2NUM(p->pst_pgrp),
      INT2NUM(p->pst_pri),
      INT2NUM(p->pst_addr),
      INT2NUM(p->pst_cpu),
      INT2NUM(p->pst_utime),
      INT2NUM(p->pst_stime),
      rb_time_new(p->pst_start,0),
      INT2NUM(p->pst_flag),
      rb_str_new2(state),
      INT2NUM(p->pst_wchan),
      INT2NUM(p->pst_procnum),
      rb_str_new2(p->pst_cmd),
      rb_str_new2(p->pst_cmd),
      INT2NUM(p->pst_time),
      INT2NUM(p->pst_cpticks),
      INT2NUM(p->pst_cptickstotal),
      INT2NUM(p->pst_fss),
      rb_float_new(p->pst_pctcpu),
      INT2NUM(p->pst_rssize),
      INT2NUM(p->pst_suid),
      INT2NUM(p->pst_shmsize),
      INT2NUM(p->pst_mmsize),
      INT2NUM(p->pst_usize),
      INT2NUM(p->pst_iosize),
      INT2NUM(p->pst_vtsize),
      INT2NUM(p->pst_vdsize),
      INT2NUM(p->pst_vssize),
      INT2NUM(p->pst_vshmsize),
      INT2NUM(p->pst_vmmsize),
      INT2NUM(p->pst_vusize),
      INT2NUM(p->pst_viosize),
      UINT2NUM(p->pst_minorfaults),
      UINT2NUM(p->pst_majorfaults),
      UINT2NUM(p->pst_nswap),
      UINT2NUM(p->pst_nsignals),
      UINT2NUM(p->pst_msgrcv),
      UINT2NUM(p->pst_msgsnd),
      INT2NUM(p->pst_maxrss),
      INT2NUM(p->pst_sid),
      INT2NUM(p->pst_schedpolicy),
      INT2NUM(p->pst_ticksleft),
      INT2NUM(p->pst_euid),
      INT2NUM(p->pst_egid),
      INT2NUM(p->pst_gid),
      INT2NUM(p->pst_sgid)
   );

   OBJ_FREEZE(v_struct);

   return v_struct;
}

/*
 * call-seq:
 *    ProcTable.ps(pid=nil)
 *    ProcTable.ps(pid=nil){ |ps| ... }
 *
 * In block form, yields a ProcTableStruct for each process entry that you
 * have rights to.  This method returns an array of ProcTableStruct's in
 * non-block form.
 *
 * If a +pid+ is provided, then only a single ProcTableStruct is yielded or
 * returned, or nil if no process information is found for that +pid+.
 */
static VALUE proctable_ps(int argc, VALUE *argv)
{
   struct pst_status pst[BURST];
   int i, count;
   int idx = 0;
   int found = 0;
   VALUE v_pid = Qnil;
   VALUE v_arr = rb_ary_new();

   rb_scan_args(argc, argv, "01", &v_pid);

   if(v_pid != Qnil){
      struct pst_status pst_single;
      int n = NUM2INT(v_pid);
      if(pstat_getproc(&pst_single, sizeof(pst_single), 0, n) > 0){
         if(rb_block_given_p()){
            rb_yield(proctable_getprocstruct(&pst_single));
            return Qnil;
         }
         else{
            return(proctable_getprocstruct(&pst_single));
         }
      }
      else{
         return Qnil; /* not found */
      }
   }

   while ((count = pstat_getproc(pst, sizeof(pst[0]), BURST, idx)) > 0){
      for (i = 0; i < count; i++){
         if( rb_block_given_p() )
            rb_yield(proctable_getprocstruct(&pst[i]));
         else
            rb_ary_push(v_arr, proctable_getprocstruct(&pst[i]) );
      }
      idx = pst[count-1].pst_idx + 1;
   }

   if(rb_block_given_p())
      return Qnil;

   return v_arr;
}

/*
 * call-seq:
 *    ProcTable.fields
 *
 * Returns an array of fields that each ProcTableStruct will contain.  This
 * may be useful if you want to know in advance what fields are available
 * without having to perform at least one read of the /proc table.
 */
static VALUE proctable_fields()
{
   int i;
   VALUE v_farray = rb_ary_new();

   for(i = 0; i < (sizeof(fields) / sizeof(fields[0])); i++)
      rb_ary_push(v_farray,rb_str_new2(fields[i]));

   return v_farray;
}

/*
 * A Ruby interface for gathering process table information.
 */
void Init_proctable()
{
   VALUE mSys, cProcTable;

   /* The Sys module serves as a namespace only */
   mSys = rb_define_module("Sys");

   /* The ProcTable class encapsulates process table information */
   cProcTable = rb_define_class_under(mSys, "ProcTable", rb_cObject);

   /* The error typically raised if any of the ProcTable methods fail */
   cProcError = rb_define_class_under(cProcTable, "Error", rb_eStandardError);

   /* Singleton methods */

   rb_define_singleton_method(cProcTable, "fields", proctable_fields, 0);
   rb_define_singleton_method(cProcTable, "ps", proctable_ps, -1);

   /* There is no constructor */
   rb_funcall(cProcTable, rb_intern("private_class_method"), 1, ID2SYM(rb_intern("new")));

   /* 1.1.5: The version of the sys-proctable library. */
   rb_define_const(cProcTable, "VERSION", rb_str_new2("1.1.5"));

   /* Structs */

   sProcStruct = rb_struct_define("ProcTableStruct",
      "comm","uid","pid","ppid","dsize","tsize","ssize","nice","ttydev","pgrp",
      "pri","addr","cpu","utime","stime","start","flag","stat","wchan",
      "procnum","cmd","cmdline","time","cpticks","cptickstotal","fss","pctcpu",
      "rssize","suid","shmsize","mmsize", "usize","iosize","vtsize","vdsize",
      "vssize","vshmsize","vmmsize","vusize","viosize","minorfaults",
      "majorfaults","nswap","nsignals","msgrcv","msgsnd","maxrss","sid",
      "schedpolicy","ticksleft","euid","egid","gid","sgid",NULL
   );
}

#ifdef __cplusplus
}
#endif
